File: problems/leetcode/1100/1110.cpp
// 1110. Delete Nodes And Return Forest
// https://leetcode.com/problems/delete-nodes-and-return-forest/

using namespace std;

#include <vector>
#include <set>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    vector<TreeNode*> delNodes(TreeNode* root, vector<int>& to_delete) {
        vector<TreeNode*> answer;

        set<int> deleteVals;
        for(auto& val: to_delete) {
            deleteVals.insert(val);
        }

        if(deleteVals.find(root->val) != deleteVals.end()) {
            traverseForRoot(answer, deleteVals, root);
        } else {
            answer.push_back(root);
            traverse(answer, deleteVals, root);
        }

        return answer;
    }

    void traverse(vector<TreeNode*>& answer, set<int>& deleteVals, TreeNode* node) {
        // Don't need to verify current node is one of answer or not.

        if(node->left) {
            if(deleteVals.find(node->left->val) != deleteVals.end()) {
                traverseForRoot(answer, deleteVals, node->left);
                node->left = nullptr;
            } else {
                traverse(answer, deleteVals, node->left);
            }
        }

        if(node->right) {
            if(deleteVals.find(node->right->val) != deleteVals.end()) {
                traverseForRoot(answer, deleteVals, node->right);
                node->right = nullptr;
            } else {
                traverse(answer, deleteVals, node->right);
            }
        }
    }

    void traverseForRoot(vector<TreeNode*>& answer, set<int>& deleteVals, TreeNode* node) {
        // Need to verify current node is one of answer or not.

        if(node->left) {
            if(deleteVals.find(node->left->val) != deleteVals.end()) {
                traverseForRoot(answer, deleteVals, node->left);
                node->left = nullptr;
            } else {
                answer.push_back(node->left);
                traverse(answer, deleteVals, node->left);
            }
        }

        if(node->right) {
            if(deleteVals.find(node->right->val) != deleteVals.end()) {
                traverseForRoot(answer, deleteVals, node->right);
                node->right = nullptr;
            } else {
                answer.push_back(node->right);
                traverse(answer, deleteVals, node->right);
            }
        }
    }
};

File: problems/leetcode/1100/1110/1110.py
# 1110. Delete Nodes And Return Forest
# https://leetcode.com/problems/delete-nodes-and-return-forest/description

from collections import deque
from typing import *

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def delNodes(self, root: Optional[TreeNode], to_delete: List[int]) -> List[TreeNode]:
        roots = []
        removals = set(to_delete)
        if root and root.val not in removals:
            roots.append(root)

        q = deque()
        if root:
            q.append(root)

        while q:
            node = q.popleft()

            if node.val in removals:
                for child in [node.left, node.right]:
                    if child and child.val not in removals:
                        roots.append(child)

            if node.left:
                q.append(node.left)
                if node.left.val in removals:
                    node.left = None

            if node.right:
                q.append(node.right)
                if node.right.val in removals:
                    node.right = None

        return roots
